src/subscribe.rs

```diff
@@ -0,0 +1,117 @@
+struct Subscriber {
+    name: Option<String>,
+    email: String,
+    phone: Option<String>,
+}
+
+impl Subscriber {
+    fn to_provider_data(&self) -> ft_sdk::auth::ProviderData {
+        ft_sdk::auth::ProviderData {
+            identity: self.email.clone(),
+            username: self.name.clone(),
+            name: self.name.clone(),
+            emails: vec![self.email.clone()],
+            verified_emails: vec![],
+            profile_picture: None,
+            custom: serde_json::json!({}),
+        }
+    }
+}
+
+/// construct [Subscriber] from request data and session
+/// authenticated user data takes precedence over request data
+fn validate(
+    conn: &mut ft_sdk::Connection,
+    name: Option<String>,
+    phone: Option<String>,
+    email: Option<String>,
+    sid: ft_sdk::Cookie<{ ft_sdk::auth::SESSION_KEY }>,
+) -> Result<(Subscriber, Option<ft_sdk::auth::UserId>), ft_sdk::Error> {
+    match ft_sdk::auth::ud(sid, conn)? {
+        Some(ud) => Ok((
+            Subscriber {
+                name: Some(ud.name),
+                email: ud.email,
+                phone,
+            },
+            Some(ft_sdk::auth::UserId(ud.id)),
+        )),
+        None => {
+            if email.is_none() {
+                return Err(ft_sdk::single_error("email", "email is required").into());
+            }
+
+            Ok((
+                Subscriber {
+                    name,
+                    email: email.unwrap(),
+                    phone,
+                },
+                None,
+            ))
+        }
+    }
+}
+
+#[ft_sdk::data]
+fn subscribe(
+    ft_sdk::Query(name): ft_sdk::Query<"name", Option<String>>,
+    ft_sdk::Query(phone): ft_sdk::Query<"phone", Option<String>>,
+    ft_sdk::Query(email): ft_sdk::Query<"email", Option<String>>,
+    ft_sdk::Query(source): ft_sdk::Query<"source", Option<String>>,
+    ft_sdk::Query(topic): ft_sdk::Query<"topic", Option<String>>,
+    sid: ft_sdk::Cookie<{ ft_sdk::auth::SESSION_KEY }>,
```

Comment:
You will need another cookie, tracker_id. We have to create `fastn_tracker` similar to `fastn_session`, but without `data`, as it will only keep track of user.

Reply:
creating a `tid` cookie is not the scope of this PR. However, the `is-subscribed` handler takes this cookie into account. The `/t/` handler will be added later via another PR